zipline/pipeline/data/dataset.py

```diff
@@ -497,3 +500,229 @@ class CompanyMetadata(DataSet):
 # base DataSet class, and we also don't want to accidentally use a shared
 # version of this attribute if we fail to set this in a subclass somewhere.
 del DataSet._domain_specializations
+
+
+class MultiDimensionalDataSetMeta(abc.ABCMeta):
+    _base_marker = object()
+
+    def __new__(cls, name, bases, dict_):
+        columns = {}
+        for k, v in dict_.copy().items():
+            if isinstance(v, Column):
+                columns[k] = v
+                del dict_[k]
+
+        is_base_class = bases == (cls._base_marker,)
+        if is_base_class:
+            bases = (object,)
+        self = super(MultiDimensionalDataSetMeta, cls).__new__(
+            cls,
+            name,
+            bases,
+            dict_,
+        )
+
+        if not is_base_class:
+            self.free_axes = free_axes = OrderedDict([
+                (k, frozenset(v))
+                for k, v in OrderedDict(self.free_axes).items()
+            ])
+            if not free_axes:
+                raise ValueError(
+                    'MultiDimensionalDataSet must be defined with non-empty'
+                    ' free_axes',
+                )
+
+            class Slice(self._SliceType):
```

Comment:
Can we call this like "BaseSlice" to help differentiate from the other `Slice` defined in the `slice` classmethod?